File: server/src/main/scala/kpn/shared/network/NetworkAttributes.scala
package kpn.shared.network

import kpn.shared.Country
import kpn.shared.LatLonImpl
import kpn.shared.NetworkType
import kpn.shared.Timestamp
import kpn.shared.common.ToStringBuilder

case class NetworkAttributes(
  id: Long,
  country: Option[Country],
  networkType: NetworkType,
  name: String,
  km: Int,
  meters: Int,
  nodeCount: Int,
  routeCount: Int,
  brokenRouteCount: Int,
  brokenRoutePercentage: String,
  integrity: Integrity,
  unaccessibleRouteCount: Int,
  connectionCount: Int,
  lastUpdated: Timestamp,
  relationLastUpdated: Timestamp,
  center: Option[LatLonImpl]
) {

  def percentageOkString: String = {
    s"${percentageOk.toInt}%"
  }

  private def percentageOk: Double = {
    if (routeCount == 0) {
      0
    }
    else {
      100d * (routeCount - brokenRouteCount) / routeCount
    }
  }

  def happy: Boolean = percentageOk > 95

  def veryHappy: Boolean = percentageOk > 99.9

  override def toString: String = ToStringBuilder(this.getClass.getSimpleName).
    field("id", id).
    field("country", country).
    field("networkType", networkType).
    field("name", name).
    field("km", km).
    field("meters", meters).
    field("nodeCount", nodeCount).
    field("routeCount", routeCount).
    field("brokenRouteCount", brokenRouteCount).
    field("brokenRoutePercentage", brokenRoutePercentage).
    field("integrity", integrity).
    field("unaccessibleRouteCount", unaccessibleRouteCount).
    field("connectionCount", connectionCount).
    field("lastUpdated", lastUpdated).
    field("relationLastUpdated", relationLastUpdated).
    field("center", center).
    build
}

File: backend/api/src/main/java/kpn/api/common/network/NetworkAttributes.java
package kpn.api.common.network;

import kpn.api.common.Country;
import kpn.api.common.LatLonImpl;
import kpn.api.common.RouteScope;
import kpn.api.common.RouteType;
import kpn.api.common.network.Integrity;
import kpn.api.custom.Timestamp;

import java.util.Optional;

public record NetworkAttributes(
  Long id,
  Optional<Country> country,
  RouteType routeType,
  RouteScope routeScope,
  Optional<String> name,
  Long km,
  Long meters,
  Long nodeCount,
  Long routeCount,
  Long brokenRouteCount,
  String brokenRoutePercentage,
  Integrity integrity,
  Long inaccessibleRouteCount,
  Long connectionCount,
  Timestamp lastUpdated,
  Timestamp relationLastUpdated,
  Optional<LatLonImpl> center
) {}

/* TODO migrate
package kpn.api.common.network

import kpn.api.common.Country
import kpn.api.common.LatLonImpl
import kpn.api.common.RouteScope
import kpn.api.common.RouteType
import kpn.api.custom.ScopedRouteType
import kpn.api.custom.Timestamp

case class NetworkAttributes(
  id: Long,
  country: Option[Country],
  routeType: RouteType,
  routeScope: RouteScope,
  name: Option[String],
  km: Long,
  meters: Long,
  nodeCount: Long,
  routeCount: Long,
  brokenRouteCount: Long,
  brokenRoutePercentage: String,
  integrity: Integrity,
  inaccessibleRouteCount: Long,
  connectionCount: Long,
  lastUpdated: Timestamp,
  relationLastUpdated: Timestamp,
  center: Option[LatLonImpl]
) {

  def scopedRouteType: ScopedRouteType = {
    ScopedRouteType.from(routeType, routeScope)
  }

  def percentageOkString: String = {
    s"${percentageOk.toInt}%"
  }

  private def percentageOk: Double = {
    if (routeCount == 0) {
      0
    }
    else {
      100d * (routeCount - brokenRouteCount) / routeCount
    }
  }

  def happy: Boolean = percentageOk > 95

  def veryHappy: Boolean = percentageOk > 99.9
}

*/
